incubator/hnc/internal/reconcilers/object.go

```diff
@@ -325,34 +327,65 @@ func (r *ObjectReconciler) syncObject(ctx context.Context, log logr.Logger, inst
 		return actionNop, nil
 	}
 
-	// This object is a propagated copy if it has "api.LabelInheritedFrom" label.
-	if hasPropagatedLabel(inst) {
+	// If the object should be propagated, we will sync it as an propagated object.
+	if r.shouldSyncAsPropagated(log, inst) {
 		return r.syncPropagated(ctx, log, inst)
 	}
 
-	// Find the source object of the same name in the ancestors from top down to
-	// see if there's a conflicting source.
-	srcInst := r.Forest.Get(inst.GetNamespace()).GetSource(r.GVK, inst.GetName())
+	r.syncSource(ctx, log, inst)
+	// No action needs to take on source objects.
+	return actionNop, nil
+}
 
-	// The object is a source without conflict if a copy of the source is not
-	// found in the forest or itself is found.
-	if srcInst == nil || srcInst.GetNamespace() == inst.GetNamespace() {
-		r.syncSource(ctx, log, inst)
-		// No action needs to take on source objects.
-		return actionNop, nil
+// shouldSyncAsPropagated checks if the object should be synced as a propagated
+// object.
+func (r *ObjectReconciler) shouldSyncAsPropagated(log logr.Logger, inst *unstructured.Unstructured) bool {
+	// This object is a propagated copy if it has "api.LabelInheritedFrom" label.
+	if hasPropagatedLabel(inst) {
+		return true
 	}
 
-	// Since there's a conflict that another source with the same name is found in
-	// the ancestors, this instance will be treated as propagated objects and will
-	// be overwritten by the source in the ancestor.
-	return r.syncPropagated(ctx, log, inst)
+	// If there's a conflicting source in the ancestors (excluding itself) and the
+	// the type has 'Propagate' mode, the object will be overwritten.
+	mode := r.Forest.GetTypeSyncer(r.GVK).GetMode()
+	if mode == api.Propagate && r.hasConflictingSource(log, inst) {
```

Comment:
Does it make sense to a) return the conflicting source here and b) return it from `shouldSyncAsPropagated` so we don't have to find this source a second time? I'm not worried about the runtime, it just seems a touch unfortunate to go through this logic twice.

But if it makes the code more convoluted, don't worry about it.

Reply:
Updated to return `srcInst`.